Add docs for `TiCDC support checking if data is entirely replicated to Downstream` (#15766)
pingcap/docs-cn: ticdc/ticdc-open-api-v2.md

响应的 JSON 格式以及字段含义与[创建同步任务](#创建同步任务)中的响应参数相同，此处不再赘述。

## 查询特定同步任务是否完成

该接口是一个同步接口，请求成功后会返回指定同步任务 (changefeed) 的同步完成情况，包括是否同步完成，以及一些更详细的信息。

### 请求 URI

`GET /api/v2/changefeed/{changefeed_id}/synced`

### 参数说明

#### 路径参数

| 参数名             | 说明                          |
|:----------------|:----------------------------|
| `changefeed_id` | 需要查询的同步任务 (changefeed) 的 ID |

### 使用样例

以下请求会查询 ID 为 `test1` 的同步任务的同步完成状态。

```shell
curl -X GET http://127.0.0.1:8300/api/v2/changefeed/test1/synced
```

> **示例 1：**

Reviewer: ```suggestion
**示例 1：同步已完成**
```

```json
{
  "synced": true,
  "sink_checkpoint_ts": "2023-11-30 15:14:11.015",
  "puller_resolved_ts": "2023-11-30 15:14:12.215",
  "last_synced_ts": "2023-11-30 15:08:35.510",
  "now_ts": "2023-11-30 15:14:11.511",
  "info": "Data syncing is finished"
}
```

以上返回的信息的说明如下：

- `synced`：该同步任务是否已完成。`true` 表示已完成；`false` 表示不一定完成，具体状态需要结合 `info` 字段以及其他字段进行判断。
- `sink_checkpoint_ts`：sink 模块的 checkpoint-ts 值，时间为 PD 时间。
- `puller_resolved_ts`：puller 模块的 resolved-ts 值，时间为 PD 时间。
- `last_synced_ts`：TiCDC 处理的最新一条数据的 commit-ts 值，时间为 PD 时间。

> - `now_ts`：当前的 PD 时间

Reviewer: ```suggestion
- `now_ts`：当前的 PD 时间。
```

> - `now_ts`: 当前的 PD 时间
> - `info`: 一些帮助判断同步状态的信息，特别是在 `synced` 为 `false` 时可以为你提供参考。

Reviewer 2: ```suggestion
- `now_ts`：当前的 PD 时间
- `info`：一些帮助判断同步状态的信息，特别是在 `synced` 为 `false` 时可以为你提供参考。
```

> **示例 2：**

Reviewer: ```suggestion
**示例 2：同步未完成**
```

```json
{
  "synced": false,
  "sink_checkpoint_ts": "2023-11-30 15:26:31.519",
  "puller_resolved_ts": "2023-11-30 15:26:23.525",
  "last_synced_ts": "2023-11-30 15:24:30.115",
  "now_ts": "2023-11-30 15:26:31.511",
  "info": "The data syncing is not finished, please wait"
}
```

此示例展示了当未完成同步任务时该接口返回的查询结果。你可以结合 `synced` 和 `info` 字段判断出数据目前还未完成同步，需要继续等待。

> **示例 3：**

Reviewer: ```suggestion
**示例 3：同步状态不确定**
```

```json
{
  "synced":false,
  "sink_checkpoint_ts":"2023-12-13 11:45:13.515",
  "puller_resolved_ts":"2023-12-13 11:45:13.525",
  "last_synced_ts":"2023-12-13 11:45:07.575",
  "now_ts":"2023-12-13 11:50:24.875",
  "info":"Please check whether PD is online and TiKV Regions are all available. If PD is offline or some TiKV regions are not available, it means that the data syncing process is complete. To check whether TiKV regions are all available, you can view 'TiKV-Details' > 'Resolved-Ts' > 'Max Leader Resolved TS gap' on Grafana. If the gap is large, such as a few minutes, it means that some regions in TiKV are unavailable. Otherwise, if the gap is small and PD is online, it means the data syncing is incomplete, so please wait"
}
```

> 本接口支持在上游集群发生灾害时对同步状态进行查询判断。在部分情况下，你可能无法直接判定 TiCDC 目前的数据同步任务是否完成。此时，你可以查询该接口，并结合返回结果中的 `info` 字段以及目前上游集群的状态进行判断。

Reviewer: ```suggestion
本接口支持在上游集群发生灾难时对同步状态进行查询判断。在部分情况下，你可能无法直接判定 TiCDC 目前的数据同步任务是否完成。此时，你可以查询该接口，并结合返回结果中的 `info` 字段以及目前上游集群的状态进行判断。
```

> 在此示例中，`sink_checkpoint_ts` 在时间上落后于 `now_ts`，这可能是因为 TiCDC 还在追数据导致的，也可能是由于 PD 或者 TiKV 出现了故障导致的。如果这是 TiCDC 还在追数据导致的，说明同步任务尚未完成。如果这是由于 PD 或者 TiKV 出现了故障导致的，说明同步任务已经完成。因此，你需要参考 `info` 中的信息对集群状态进行辅助判断。

Reviewer: ```suggestion
在此示例中，`sink_checkpoint_ts` 在时间上落后于 `now_ts`，这可能是因为 TiCDC 还在追数据，也可能是由于 PD 或者 TiKV 出现了故障。如果这是 TiCDC 还在追数据导致的，说明同步任务尚未完成。如果这是由于 PD 或者 TiKV 出现了故障导致的，说明同步任务已经完成。因此，你需要参考 `info` 中的信息对集群状态进行辅助判断。
```

> **示例 4：**

Reviewer: ```suggestion
**示例 4：查询错误**
```

```json
{
  "error_msg": "[CDC:ErrPDEtcdAPIError]etcd api call error: context deadline exceeded",
  "error_code": "CDC:ErrPDEtcdAPIError"
}
```

> 当上游集群的 PD 长时间故障后，查询该 API 接口会返回类似如上的错误，无法提供进一步的判断信息。因为 PD 故障本身会影响 TiCDC 的数据同步，当遇到该错误时，你可以认为 TiCDC 已经尽可能完成数据同步，但下游集群仍然可能存在因 PD 故障导致的数据丢失。

Reviewer: ```suggestion
在上游集群的 PD 长时间故障后，查询该 API 接口时会返回类似如上的错误，无法提供进一步的判断信息。因为 PD 故障会直接影响 TiCDC 的数据同步，当遇到这类错误时，你可以认为 TiCDC 已经尽可能完成数据同步，但下游集群仍然可能存在因 PD 故障导致的数据丢失。
```

```

如果请求成功，则返回 `200 OK`。如果请求失败，则返回错误信息和错误码。

## 查询特定同步任务同步是否完成

该接口是一个同步接口，请求成功后会返回指定的同步任务(changefeed)的同步完成情况，包括是否同步完成，以及一些更详细的信息。

### 请求 URI

`GET /api/v2/changefeed/{changefeed_id}/synced`

### 参数说明

#### 路径参数

| 参数名             | 说明                          |
|:----------------|:----------------------------|
| `changefeed_id` | 需要查询的同步任务 (changefeed) 的 ID |

### 使用样例

以下请求会查询 ID 为 `test1` 的同步任务的同步完成状态。

```shell
curl -X GET http://127.0.0.1:8300/api/v2/changefeed/test1/synced
```

示例1

```json
{
  "synced": true,
  "sink_checkpoint_ts": "2023-11-30 15:14:11",
  "puller_resolved_ts": "2023-11-30 15:14:09",
  "last_synced_ts": "2023-11-30 15:08:35",
  "now_ts": "2023-11-30 15:14:11",
  "info": "Data syncing is finished"
}
```

以上返回的信息的说明如下：

- `synced`：该同步任务是否已经彻底同步完成了，true 则表示一定完成了同步任务，false 则表示并不一定完成了同步任务，具体需要结合 "info" 字段描述以及其他字段进行判断。
- `sink_checkpoint_ts`: sink 模块的 checkpoint-ts 值，时间为 PD 时间。
- `puller_resolved_ts`: puller 模块的 resolved-ts 值，时间为 PD 时间。

> - `last_synced_ts`: ticdc 处理的最新一条数据的 commit-ts 值，时间为 PD 时间。

Reviewer 3: 这应该是指 syncpint 吧？https://docs.pingcap.com/zh/tidb/stable/ticdc-upstream-downstream-check#%E5%90%AF%E7%94%A8-syncpoint

那这是下游时间还是上游时间？

Author: 这个不是 syncpoint，这个是这个 feature 新加的一个字段。都是上游 PD 算的时间

Author: last-synced-ts 可以认为就是 ticdc 最后处理的那条 event 的 commitTS

- `now_ts`: 当前的 PD 时间
- `info`: 一些帮助判断的信息，主要在 synced 为 false 时候使用。

示例2

```json
{
  "synced": false,
  "sink_checkpoint_ts": "2023-11-30 15:26:31",
  "puller_resolved_ts": "2023-11-30 15:26:23",
  "last_synced_ts": "2023-11-30 15:24:30",
  "now_ts": "2023-11-30 15:26:31",
  "info": "The data syncing is not finished, please wait"
}

{
  "synced":false,
  "sink_checkpoint_ts":"2023-12-13 11:45:13",
  "puller_resolved_ts":"2023-12-13 11:45:13",
  "last_synced_ts":"2023-12-13 11:45:07",
  "now_ts":"2023-12-13 11:47:24",

>   "info":"Please check whether pd is health and tikv region is all available. If pd is not health or tikv region is not available, the data syncing is finished.  Otherwise the data syncing is not finished, please wait"

Reviewer 3: “the data syncing is finished” 后面补一下判断条件

Author: 这个是说我需要解释一下为什么 pd 或者 tikv unavailable 的时候，我们判断 data syncing is finished 嘛？

Author: 我这边补充了一些解释，麻烦看看这样可以么

>   "info":"Please check whether pd is health and tikv region is all available. If pd is not health or tikv region is not available, the data syncing is finished.  Otherwise the data syncing is not finished, please wait"

Reviewer 4: I would suggest changing "health" to "healthy" in this text as it is being used as an adjective.

Author: Good Point! I will change it to "healthy"

}

{
  "error_msg": "[CDC:ErrPDEtcdAPIError]etcd api call error: context deadline exceeded",
  "error_code": "CDC:ErrPDEtcdAPIError"
}
```

> 该示例展示的是并未严格的完成同步任务时返回的查询结果，我们可以结合 `synced` 和 `info` 字段确认数据目前同步的状态。

Reviewer 3: 建议在每一个示例后面分别加一个解释。上面分成 3 个示例

```

如果请求成功，则返回 `200 OK`。如果请求失败，则返回错误信息和错误码。

## 查询特定同步任务同步是否完成

> 该接口是一个同步接口，请求成功后会返回指定的同步任务(changefeed)的同步完成情况，包括是否同步完成，以及一些更详细的信息。

Reviewer 2: ```suggestion
该接口是一个同步接口，请求成功后会返回指定同步任务 (changefeed) 的同步完成情况，包括是否同步完成，以及一些更详细的信息。
```

### 请求 URI

`GET /api/v2/changefeed/{changefeed_id}/synced`

### 参数说明

#### 路径参数

| 参数名             | 说明                          |
|:----------------|:----------------------------|
| `changefeed_id` | 需要查询的同步任务 (changefeed) 的 ID |

### 使用样例

以下请求会查询 ID 为 `test1` 的同步任务的同步完成状态。

```shell
curl -X GET http://127.0.0.1:8300/api/v2/changefeed/test1/synced
```

> 示例1

Reviewer 2: ```suggestion
**示例 1：**
```

```json
{
  "synced": true,
  "sink_checkpoint_ts": "2023-11-30 15:14:11.015",
  "puller_resolved_ts": "2023-11-30 15:14:12.215",
  "last_synced_ts": "2023-11-30 15:08:35.510",
  "now_ts": "2023-11-30 15:14:11.511",
  "info": "Data syncing is finished"
}
```

以上返回的信息的说明如下：

> - `synced`：该同步任务是否已经彻底同步完成了，true 则表示一定完成了同步任务，false 则表示并不一定完成了同步任务，具体需要结合 "info" 字段描述以及其他字段进行判断。

Reviewer 2: ```suggestion
- `synced`：该同步任务是否已完成。`true` 表示确定已完成；`false` 表示不一定完成，具体状态需要结合 `info` 字段以及其他字段进行判断。
```

> - `sink_checkpoint_ts`: sink 模块的 checkpoint-ts 值，时间为 PD 时间。
> - `puller_resolved_ts`: puller 模块的 resolved-ts 值，时间为 PD 时间。
> - `last_synced_ts`: ticdc 处理的最新一条数据的 commit-ts 值，时间为 PD 时间。

Reviewer 2: ```suggestion
- `sink_checkpoint_ts`：sink 模块的 checkpoint-ts 值，时间为 PD 时间。
- `puller_resolved_ts`：puller 模块的 resolved-ts 值，时间为 PD 时间。
- `last_synced_ts`：TiCDC 处理的最新一条数据的 commit-ts 值，时间为 PD 时间。
```

- `now_ts`: 当前的 PD 时间

> - `info`: 一些帮助判断的信息，主要在 synced 为 false 时候使用。

Reviewer 2: ```suggestion
- `info`: 一些帮助判断同步状态的信息，特别是在 `synced` 为 `false` 时可以为你提供参考。
```

> 示例2

Reviewer 2: ```suggestion
**示例 2：**
```

```json
{
  "synced": false,
  "sink_checkpoint_ts": "2023-11-30 15:26:31.519",
  "puller_resolved_ts": "2023-11-30 15:26:23.525",
  "last_synced_ts": "2023-11-30 15:24:30.115",
  "now_ts": "2023-11-30 15:26:31.511",
  "info": "The data syncing is not finished, please wait"
}
```

> 示例2 展示的是并未完成同步任务时返回的查询结果，我们可以结合 `synced` 和 `info` 字段确认数据目前还未完成同步，需要继续等待。

Reviewer 2: ```suggestion
此示例展示了当未完成同步任务时该接口返回的查询结果。你可以结合 `synced` 和 `info` 字段确认数据目前还未完成同步，需要继续等待。
```

> 示例3

Reviewer 2: ```suggestion
**示例 3：**
```

```json
{
  "synced":false,
  "sink_checkpoint_ts":"2023-12-13 11:45:13.515",
  "puller_resolved_ts":"2023-12-13 11:45:13.525",
  "last_synced_ts":"2023-12-13 11:45:07.575",
  "now_ts":"2023-12-13 11:50:24.875",

>   "info":"Please check whether pd is healthy and tikv region is all available. If pd is not healthy or tikv region is not available, the data syncing is finished. When pd is offline means that pd is not healthy. For tikv region, you can check the grafana info in 'TiKV-Details-Resolved-Ts-Max Leader Resolved TS gap'. If the gap is a large value, such as a few minutes, it means some regions in tikv are unavailable. Otherwise the data syncing is not finished, please wait"

Reviewer 2: ```suggestion
  "info":"Please check whether PD is online and TiKV Regions are all available. If PD is offline or TiKV regions are not available, it means that the data syncing process is complete. To learn whether TiKV regions are all available, you can check the 'TiKV-Details-Resolved-Ts-Max Leader Resolved TS gap' metric on Grafana. If the gap is large, such as a few minutes, it means that some regions in TiKV are unavailable. Otherwise, if the gap is small and PD is online, it means the data syncing is incomplete, so please wait"
```

}
```

> 因为本接口支持在上游集群发生灾害的时候进行查询判断，因此在部分情况下，我们无法直接判定 TiCDC 目前数据同步是否完成，而需要用户根据 `info` 信息以及目前上游集群的状态进行判断。在这个示例中 `sink_checkpoint_ts` 和 `now_ts` 具有一些时间上的差距，但我们不知道上游集群的状态，所以不能判定是 TiCDC 还在追数据，所以 `checkpoint-ts` 落后, 还是 pd 或者 tikv 出现了故障导致了 `checkpoint-ts` 没有正常推进。因此我们需要用户根据集群状态来进行辅助判断。

Reviewer 3: "If pd is not healthy or tikv region is not available, the data syncing is finished." 这个结论需要一个判断依据，即告诉用户具体应该怎么判断。

Author: Done， 麻烦看看现在的描述是否清晰。

> 因为本接口支持在上游集群发生灾害的时候进行查询判断，因此在部分情况下，我们无法直接判定 TiCDC 目前数据同步是否完成，而需要用户根据 `info` 信息以及目前上游集群的状态进行判断。在这个示例中 `sink_checkpoint_ts` 和 `now_ts` 具有一些时间上的差距，但我们不知道上游集群的状态，所以不能判定是 TiCDC 还在追数据，所以 `checkpoint-ts` 落后, 还是 pd 或者 tikv 出现了故障导致了 `checkpoint-ts` 没有正常推进。因此我们需要用户根据集群状态来进行辅助判断。

Reviewer 3: ```suggestion
因为本接口支持在上游集群发生灾害的时候进行查询判断，因此在部分情况下，我们无法直接判定 TiCDC 目前数据同步是否完成，而需要用户根据 `info` 信息以及目前上游集群的状态进行判断。在这个示例中 `sink_checkpoint_ts` 和 `now_ts` 具有一些时间上的差距，但我们不知道上游集群的状态，所以不能判定是 TiCDC 还在追数据导致 `checkpoint-ts` 落后, 还是 pd 或者 tikv 出现了故障导致了 `checkpoint-ts` 没有正常推进。因此我们需要用户根据集群状态来进行辅助判断。
```

> 因为本接口支持在上游集群发生灾害的时候进行查询判断，因此在部分情况下，我们无法直接判定 TiCDC 目前数据同步是否完成，而需要用户根据 `info` 信息以及目前上游集群的状态进行判断。在这个示例中 `sink_checkpoint_ts` 和 `now_ts` 具有一些时间上的差距，但我们不知道上游集群的状态，所以不能判定是 TiCDC 还在追数据导致 `checkpoint-ts` 落后, 还是 pd 或者 tikv 出现了故障导致了 `checkpoint-ts` 没有正常推进。因此我们需要用户根据集群状态来进行辅助判断。

Reviewer 2: ```suggestion
本接口支持在上游集群发生灾害的时候查询判断同步状态。在部分情况下，你可能无法直接判定 TiCDC 目前的数据同步任务是否完成。此时，你可以查询该接口，并结合返回结果中 `info` 信息以及目前上游集群的状态进行判断。在此示例中，`sink_checkpoint_ts` 和 `now_ts` 之间存在一些时间上的差距，但上游集群的状态未知，所以无法直接确定 `sink_checkpoint_ts` 落后的原因，是由于 TiCDC 还在追数据 ，还是由于 PD 或者 TiKV 出现了故障导致 `checkpoint-ts` 没有正常推进。因此需要你参考 `info` 中的信息对集群状态进行辅助判断。
```

> 示例4

Reviewer 2: ```suggestion
**示例 4：**
```

```json
{
  "error_msg": "[CDC:ErrPDEtcdAPIError]etcd api call error: context deadline exceeded",
  "error_code": "CDC:ErrPDEtcdAPIError"
}
```

> 当上游集群的 PD 长时间故障后，API 查询会直接返回类似如上的错误，无法提供进一步的判断信息。而因为 PD 故障本身会影响 TiCDC 的数据推进，因此用户可以认为 TiCDC 已经尽可能完成数据同步。

Reviewer 2: ```suggestion
当上游集群的 PD 长时间故障后，查询该 API 接口会直接返回类似如上的错误，无法提供进一步的判断信息。这是因为 PD 故障本身会影响 TiCDC 的数据推进，因此你可以认为 TiCDC 已经尽可能完成数据同步。
```
